### DISCUSSION TITLE
How to constrain n parameters to be nonzero?

### DISCUSSION THREAD

**Discussion by SebBT01** [meta-pytorch/botorch#2000] [General]
Hey Botorch community,

I want to specify a search space in which there are e.g. 20 parameters but I want to constrain the optimization to contain at minimum e.g. 6. While including all 20 is technically valid, it will be useful for the optimizer if it can drop some of them and focus on a selection. I can specify the parameters as continuous between 0 and 1. 

I haven't been able to think of a way to do this.. any ideas?

EDIT: Appears a solution would be an outcome constraint with a callable checking that e.g. at minimum 6 parameters are nonzero and outputting 1 if thats the case or 0?

Cheers

**Comment by Balandat** [comment]
> EDIT: Appears a solution would be an outcome constraint with a callable checking that e.g. at minimum 6 parameters are nonzero and outputting 1 if thats the case or 0?

This may be hard from a modeling perspective. Better to address this without adding additional outcomes and constraints on those. 

This work on Sparse BO is highly relevant: https://arxiv.org/pdf/2203.01900.pdf. This Ax tutorial shows how to use this: https://github.com/facebook/Ax/blob/6443cee30cbf8cec290200a7420a3db08e4b5445/tutorials/sebo.ipynb - if you can use that directly in Ax, great. If you need it in BoTorch you should be able to just pull the `SEBOAcquisition` function from the Ax repo and mimic what the tutorial does. 

cc @dme65, @qingfeng10

**Reply by SebBT01** [reply]
Thank you for the advice! I have tried it out with simulated data, and it does indeed look like it behaves like desired. The fact that this was published a mere 6 months ago, has a great implementation and tutorial is whats so great about open source software and research.
